packages/core/src/components/dialog/dialog.tsx

```diff
@@ -121,14 +126,19 @@ export class Dialog extends AbstractPureComponent2<DialogProps> {
             <Overlay {...this.props} className={Classes.OVERLAY_SCROLL_CONTAINER} hasBackdrop={true}>
                 <div className={Classes.DIALOG_CONTAINER} ref={this.props.containerRef}>
                     <div
-                        className={classNames(Classes.DIALOG, this.props.className)}
+                        className={classNames(Classes.DIALOG, this.props.footer != null && Classes.DIALOG_NO_PADDING, this.props.className)}
```

Comment:
classNames has a slightly nicer syntax for conditional classes:

```suggestion
                        className={classNames(Classes.DIALOG, this.props.className, {
                            [Classes.DIALOG_NO_PADDING]: this.props.footer != null,
                        })}
```